modules/fluid_properties/include/utils/NewtonInversion.h

```diff
@@ -23,40 +24,181 @@ namespace FluidPropertiesUtils
  * convergence
  * @param[in] function two-variable function returning both values and derivatives as references
  * @param[in] max_its the maximum number of iterations for Newton's method
- * @return the value z such that f(x, z) = y
+ * @return a pair in which the first member is the value z such that f(x, z) = y and the second
+ * member is dy/dz
  */
-Real NewtonSolve(const Real & x,
-                 const Real & y,
-                 const Real & z_initial_guess,
-                 const Real & tolerance,
-                 std::function<void(Real, Real, Real &, Real &, Real &)> const & func,
-                 const unsigned int max_its = 100);
+template <typename T, typename Functor>
+std::pair<T, T>
+NewtonSolve(const T & x,
+            const T & y,
+            const Real z_initial_guess,
+            const Real tolerance,
+            const Functor & func,
+            const unsigned int max_its = 100)
+{
+  std::function<bool(const T &, const T &)> abs_tol_check =
+      [tolerance](const T & f, const T & /*y*/)
+  { return MetaPhysicL::raw_value(std::abs(f)) < tolerance; };
+  std::function<bool(const T &, const T &)> rel_tol_check = [tolerance](const T & f, const T & y)
+  { return MetaPhysicL::raw_value(std::abs(f / y)) < tolerance; };
+  auto convergence_check = MetaPhysicL::raw_value(y) == 0 ? abs_tol_check : rel_tol_check;
```

Comment:
But my desire is that we use a relative convergence check when we can

Reply:
I know, but supposing y can be any value (and I haven't thought about if this would ever/often be the case), you'll always fail to converge if y is very small but not zero. For example, supposing your tolerance is 1e-8 and y is 1e-12, you'd be requiring that f be < 1e-20.

Reply:
that's a good point